internal/task: use a lock-free queue

niaow opened the pull request:
This changes task.Stack and task.Queue to use atomic compare-and-swap operations.
With this change, Push can also be executed safely from multiple threads/cores.
Additionally, the queue is actually LIFO now (not sure what I was thinking before).

Not entirely sure about this yet.

niaow commented:
Oops, i meant FIFO not LIFO. Also xtensa seems to still need work?

niaow commented:
Size impact statistics:
- Arduino Uno Blinky (tasks): adds 172 bytes
- Arduino Nano 33 Blinky: adds 84 bytes
- Metro M4 Airlift Blinky: adds 40 bytes
- ESP32-Mini32 Blinky: adds 40 bytes
- Linux Channel Test: adds 76 bytes

aykevl commented on src/runtime/runtime_esp8266.go:
```diff
@@ -115,3 +117,23 @@ func abort() {
 }
+
+//export __sync_val_compare_and_swap_4
+func __sync_val_compare_and_swap_4(ptr *uint32, expected, desired uint32) uint32 {
```
Huh, interesting. These functions haven't been implemented before:

```
$ build/tinygo build -o test.elf -target=esp8266 ./testdata/atomic.go
/tmp/tinygo787979806/main.o:(.literal.runtime.run$1$gowrapper+0x0): undefined reference to `__sync_fetch_and_add_4'
/tmp/tinygo787979806/main.o:(.literal.runtime.run$1$gowrapper+0x4): undefined reference to `__sync_fetch_and_add_8'
/tmp/tinygo787979806/main.o:(.literal.runtime.run$1$gowrapper+0x8): undefined reference to `__sync_lock_test_and_set_4'
/tmp/tinygo787979806/main.o:(.literal.runtime.run$1$gowrapper+0xc): undefined reference to `__sync_lock_test_and_set_8'
/tmp/tinygo787979806/main.o:(.literal.runtime.run$1$gowrapper+0x10): undefined reference to `__sync_val_compare_and_swap_4'
/tmp/tinygo787979806/main.o:(.literal.runtime.run$1$gowrapper+0x14): undefined reference to `__sync_val_compare_and_swap_8'
/tmp/tinygo787979806/main.o:(.literal.main.testValue+0x0): undefined reference to `sync/atomic.runtime_procPin'
/tmp/tinygo787979806/main.o:(.literal.main.testValue+0x4): undefined reference to `sync/atomic.runtime_procUnpin'
error: failed to link /tmp/tinygo787979806/main: exit status 1
```

If that isn't too difficult, it would be great to have these implemented in a separate PR. Perhaps as a generalization of the functions in src/runtime/arch_cortexm.go - after all, this can be implemented the same way on all unicore processors (which are in effect all MCUs TinyGo supports as TinyGo doesn't use more than one core). I think src/runtime/baremetal.go (or even a new src/runtime/atomic.go) would be a good place for these functions.

niaow replied:
Yeah, I was planning on doing something like that soon.

niaow replied:
Moved into #2307

aykevl commented on src/runtime/arch_avr.go:
```diff
@@ -14,3 +17,14 @@ func align(ptr uintptr) uintptr {
 func getCurrentStackPointer() uintptr
+
+//export __sync_val_compare_and_swap_2
+func __sync_val_compare_and_swap_2(ptr *uint32, expected, desired uint32) uint32 {
```
The same goes for AVR.

Note that atomics are somewhat broken on AVR. If you see weird behavior on AVR, it could be the backend that is emitting invalid instructions.
For more details, see: https://reviews.llvm.org/D97127

aykevl commented on src/runtime/runtime_avr.go:
```diff
@@ -93,3 +95,12 @@ func abort() {
 }
+
+//export __sync_lock_test_and_set_2
+func __sync_lock_test_and_set_2(ptr *uint16, value uint16) uint16 {
```
Whatever the exact location, these two functions should _definitely_ be together in the same file.

niaow replied:
Yeah i think this was a mistake

aykevl reviewed (commented):
Out of curiosity: do you have plans to use the changes in this PR? Any new features this will add?

I haven't looked too deeply into the queue code but I don't see anything surprising with a quick glance.

niaow commented:
> Out of curiosity: do you have plans to use the changes in this PR?

I am not intending to do anything specific right now, but this might be useful once we want to implement multicore support for something.

> Any new features this will add?

Right now:
1. The queue. . .  is actually a queue instead of a weird mostly-stack
2. On devices with support for atomics, this does not need any critical sections

The Push operation can also be safely called concurrently. . .  but only if the tasks are pinned safely (which is currently an issue, but this is more of a future thing).

niaow committed:
internal/task: use a lock-free queue

This changes task.Stack and task.Queue to use atomic compare-and-swap operations.
With this change, they can also be executed safely from multiple threads/cores.
Additionally, the queue is actually LIFO now (not sure what I was thinking before).

niaow commented:
This has been rebased now that #2307 is merged

niaow commented:
I still need to test this fully on AVR though

niaow commented:
Now that #2458 and #2459 are in, I consider this ready.

deadprogram commented:
Seems like this is perhaps not working correctly on the HiFive1b? I tried several times, and also ran some other test code successfully.

jclab-joseph commented:
@aykevl If this is correct, would you be willing to merge it?
I'm going to try implementing threads again.

aykevl commented:
I think we can close this now?
I don't see what features it adds or what bugs it fixes, while it does increase binary size.
